homeassistant/components/monzo/__init__.py

```diff
@@ -0,0 +1,328 @@
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    ATTR_DEVICE_ID,
+    CONF_AUTHENTICATION,
+    EVENT_HOMEASSISTANT_STARTED,
+    EVENT_HOMEASSISTANT_STOP,
+    Platform,
+)
+from homeassistant.core import CoreState, Event, HomeAssistant, ServiceCall, callback
+from homeassistant.helpers import (
+    aiohttp_client,
+    config_entry_oauth2_flow,
+    device_registry as dr,
+)
+from homeassistant.helpers.device_registry import DeviceEntry, DeviceRegistry
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.event import async_call_later
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .api import AsyncConfigEntryAuth
+from .const import (
+    ACCOUNTS,
+    ATTR_AMOUNT,
+    CONF_CLOUDHOOK_URL,
+    CONF_COORDINATOR,
+    DEFAULT_AMOUNT,
+    DEPOSIT,
+    DOMAIN,
+    MODEL_POT,
+    POTS,
+    SERVICE_POT_TRANSFER,
+    SERVICE_REGISTER_WEBHOOK,
+    SERVICE_UNREGISTER_WEBHOOK,
+    TRANSFER_ACCOUNTS,
+    TRANSFER_TYPE,
+    VALID_POT_TRANSFER_ACCOUNTS,
+    WEBHOOK_DEACTIVATION,
+    WEBHOOK_PUSH_TYPE,
+)
+from .webhook import async_handle_webhook
+
+PLATFORMS: list[Platform] = [Platform.SENSOR]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Monzo from a config entry."""
+    implementation = (
+        await config_entry_oauth2_flow.async_get_config_entry_implementation(
+            hass, entry
+        )
+    )
+
+    async def async_get_monzo_api_data() -> dict[str, Any]:
+        accounts = await externalapi.user_account.accounts()
+        pots = await externalapi.user_account.pots()
+        hass.data[DOMAIN][entry.entry_id][ACCOUNTS] = accounts
+        hass.data[DOMAIN][entry.entry_id][POTS] = pots
+        return {ACCOUNTS: accounts, POTS: pots}
+
+    @callback
+    async def handle_pot_transfer(call: ServiceCall) -> None:
+        device_registry = dr.async_get(hass)
+        amount = int(call.data.get(ATTR_AMOUNT, DEFAULT_AMOUNT) * 100)
+        transfer_func: Callable[[str, str, int], Awaitable[bool]] = (
+            externalapi.user_account.pot_deposit
+            if call.data[TRANSFER_TYPE] == DEPOSIT
+            else externalapi.user_account.pot_withdraw
+        )
+
+        try:
+            account_ids, pot_ids = await _get_transfer_ids(
+                call, device_registry, hass, entry
+            )
+        except ValueError:
+            return
+
+        success = all(
+            await asyncio.gather(
+                *[
+                    transfer_func(account, pot, amount)
+                    for account in account_ids
+                    for pot in pot_ids
+                ]
+            )
+        )
+
+        if not success:
+            _LOGGER.error(
+                "External error handling pot transfer: one or more transactions failed"
+            )
+
+    async def register_webhook(
```

Comment:
Can you maybe explain a bit on how the webhooks work? What data do we get from them?

Reply:
The webhooks are used for the event triggers. When registered, Monzo will hit HA with a `transaction.created` event which can be used to trigger automations and contains lots of extra data you can use such as `amount` and `merchant`. A full list of the data provided can be found in [Monzo's docs](https://docs.monzo.com/#list-webhooks)